pgmax/fg/groups.py

```diff
@@ -359,18 +359,47 @@ class FactorGroup:
         factor_type: Factor type shared by all the Factors in the FactorGroup.
         factor_configs: Optional array containing an explicit enumeration of all valid configurations
         log_potentials: Array of log potentials.
+
+    Attributes:
+        factor_sizes: Array of the different factor sizes.
+        variables_for_factors: Tuple concatenating the variables connected to each factor in the FactorGroup.
+            Each variable will appear once for each Factor it connects to.
+        factor_edges_num_states: Array concatenating the number of states for the variables connected to each Factor of
+            the FactorGroup. Each variable will appear once for each Factor it connects to.
+
+    Raises:
+        ValueError: if the FactorGroup does not contain a Factor
     """
 
     variable_group: Union[CompositeVariableGroup, VariableGroup]
     variable_names_for_factors: Sequence[List]
     factor_type: Type = field(init=False)
     factor_configs: np.ndarray = field(init=False)
     log_potentials: np.ndarray = field(init=False, default=np.empty((0,)))
+    factor_sizes: np.ndarray = field(init=False)
+    variables_for_factors: Tuple[nodes.Variable, ...] = field(init=False)
```

Comment:
I think making `variables_for_factors` a tuple of tuples (a parallel of `variable_names_for_factors`) is cleaner, and allows us to get rid of `factor_sizes`. Would probably also come in handy in anticipation to our variables-related refactors.

Reply:
slower in general, or slower for numba?

Reply:
@wlehrach this is not related to numba but to the number of lists manipulated

![Screen Shot 2022-04-06 at 11 45 16 AM](https://user-images.githubusercontent.com/13166200/162046390-e9bdf467-0a50-44ef-9f8d-db1da6df1517.png)